jax/_src/nn/functions.py

```diff
@@ -1210,81 +1210,184 @@ def _check_shape_and_dtype(t: Array | None, shape: Sequence[int],
-        lhs_scales (Array): A 3D array of shape (B, M, K_block).
-        rhs_scales (Array): A 3D array of shape (B, N, K_block).
-        preferred_element_type (DTypeLike, optional): The preferred data type
-          for the computation. Defaults to `jnp.float32`.
+      a (Array): Shape (B, M, K).
+      b (Array): Shape (B, N, K).
+      a_scales (Array): Shape (B, M, K_a), where `K % K_a == 0`.
+      b_scales (Array): Shape (B, N, K_b), where `K % K_b == 0`.
+      preferred_element_type (DTypeLike, optional): Defaults to `jnp.float32`.
+
     Returns:
-        Array: A 3D array of shape (B, M, N) representing the scaled matrix
-          multiplication result.
-    Raises:
-        AssertionError: If the number of columns in `lhs` (`lhs_K`) does not
-          match the number of columns in `rhs` (`rhs_K`).
+      Array of shape (B, M, N).
+
     Notes:
-        - The function ensures that the `preferred_element_type` is
-          danonicalized before passing it to the underlying computation.
-        - Scaling is applied to the matrices based on the `lhs_scales` and
-          `rhs_scales` arrays, enabling efficient computations in blocks.
+      - We currently do not support user-defined `precision` for customizing the
+        compute data type. It is fixed to `jnp.float32`.
+      - Block size is inferred as `K // K_a` for `a` and `K // K_b` for `b`.
+      - To use cuDNN with Nvidia Blackwell GPUs, inputs must match::
+
+          # mxfp8
+          a, b: jnp.float8_e4m3fn | jnp.float8_e5m2
+          a_scales, b_scales: jnp.float8_e8m0fnu
+          block_size: 32
+          # nvfp4
+          a, b: jnp.float4_e2m1fn
+          a_scales, b_scales: jnp.float8_e4m3fn
+          block_size: 16
+
+    Examples:
+
+      Basic case:
+
+      >>> a = jnp.array([1, 2, 3]).reshape((1, 1, 3))  
+      >>> b = jnp.array([4, 5, 6]).reshape((1, 1, 3))  
+      >>> a_scales = jnp.array([0.5]).reshape((1, 1, 1))  
+      >>> b_scales = jnp.array([0.5]).reshape((1, 1, 1))  
+      >>> scaled_matmul(a, b, a_scales, b_scales)  
+      Array([[[8.]]], dtype=float32)  
+    
+      Using fused cuDNN call on Blackwell GPUs:
+
+      >>> a = random.normal(keys[0], (3, 128, 64), dtype=jnp.float8_e4m3fn)  
+      >>> b = random.normal(keys[1], (3, 128, 64), dtype=jnp.float8_e4m3fn)  
+      >>> a_scales = jnp.ones((3, 128, 4), dtype=jnp.float8_e8m0fnu)  
+      >>> b_scales = jnp.ones((3, 128, 4), dtype=jnp.float8_e8m0fnu)  
+      >>> scaled_matmul(a, b, a_scales, b_scales)     
     """
-    B, M, lhs_K = lhs.shape
-    _, N, rhs_K = rhs.shape
-    assert lhs_K == rhs_K
-    _, _, K_block = lhs_scales.shape
+    assert all(x.ndim == 3 for x in (a, b, a_scales, b_scales))
+    B_a, M_a, K_a = a.shape
+    B_b, N_b, K_b = b.shape
+    assert K_a == K_b and B_a == B_b
+    B_as, M_as, K_as = a_scales.shape
+    B_bs, N_bs, K_bs = b_scales.shape
+    assert K_as == K_bs and B_as == B_bs
+    assert M_as == M_a and N_bs == N_b
```

Comment:
I think assertions should be used for internal checks, not validating user input. They can even be disabled with `python -O`.

Can you change these to be exceptions with good error messages? Like, `if K_a != K_b or B_a != B_b: raise ValueError("scaled_matmul requires ...")`?

Reply:
Fixed. PTAL.